File: tools/lib/image_error.py
"""Shared normalization for production-image errors and spatial residuals."""

from __future__ import annotations

import numpy as np


def image_error_terms(
    image: np.ndarray,
    reference: np.ndarray,
) -> tuple[np.ndarray, float]:
    """Returns the four-component sum of absolute differences and the shared reference total I."""
    value = np.asarray(image, dtype=np.float64)
    target = np.asarray(reference, dtype=np.float64)
    if value.shape != target.shape or value.ndim < 1 or value.shape[0] != 4:
        raise ValueError("Stokes images must have equal shape (4, ...).")
    numerator = np.sum(
        np.abs(value - target),
        axis=tuple(range(1, value.ndim)),
    )
    denominator = float(np.sum(np.abs(target[0])))
    if (
        not np.all(np.isfinite(numerator))
        or not np.isfinite(denominator)
        or denominator <= 0.0
    ):
        raise ValueError(
            "Stokes image error terms require finite values and positive reference I.")
    return numerator, denominator


def image_error(image: np.ndarray, reference: np.ndarray) -> np.ndarray:
    """Use the reference image's total Stokes I as the shared $L_1$ denominator for I/Q/U/V."""
    numerator, denominator = image_error_terms(image, reference)
    result = numerator / denominator
    if not np.all(np.isfinite(result)):
        raise ValueError("Stokes image error is non-finite.")
    return result


def temporal_mean_std(values: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    """Calculate the arithmetic mean and sample standard deviation of frame-by-frame errors along the first axis."""
    errors = np.asarray(values, dtype=np.float64)
    if errors.ndim == 0 or errors.shape[0] == 0:
        raise ValueError("Temporal error statistics require at least one frame.")
    if not np.all(np.isfinite(errors)) or np.any(errors < 0.0):
        raise ValueError("Temporal error statistics require finite non-negative values.")
    mean = np.mean(errors, axis=0)
    if errors.shape[0] == 1:
        std = np.full_like(mean, np.nan, dtype=np.float64)
    else:
        std = np.std(errors, axis=0, ddof=1)
    return mean, std


def peak_i_residual(image: np.ndarray, reference: np.ndarray) -> np.ndarray:
    """Pixel-wise spatial residuals are normalized by the peak Stokes I of the entire reference image."""
    value = np.asarray(image, dtype=np.float64)
    target = np.asarray(reference, dtype=np.float64)
    if value.shape != target.shape or value.ndim < 1 or value.shape[0] != 4:
        raise ValueError("Stokes images must have equal shape (4, ...).")
    denominator = float(np.max(np.abs(target[0])))
    if not np.isfinite(denominator) or denominator <= 0.0:
        raise ValueError("Reference peak Stokes I must be finite and positive.")
    result = (value - target) / denominator
    if not np.all(np.isfinite(result)):
        raise ValueError("Stokes spatial residual is non-finite.")
    return result

File: src/grrt/ImageError.cpp
#include "ImageError.h"

#include <cmath>
#include <stdexcept>
#include <string>

namespace grrt {

namespace {

void require_finite_stokes(
    std::span<const StokesPixel> image,
    const char* role) {

    constexpr std::array<const char*, 4> names = {"I", "Q", "U", "V"};
    for (size_t pixel = 0; pixel < image.size(); pixel++) {
        for (size_t stokes = 0; stokes < names.size(); stokes++) {
            if (!std::isfinite(image[pixel][stokes])) {
                throw std::domain_error(
                    std::string("Image error ") + role +
                    " has non-finite Stokes " + names[stokes] +
                    " at pixel " + std::to_string(pixel) + ".");
            }
        }
    }
}

} // namespace

std::array<double, 4> image_error(
    std::span<const StokesPixel> image,
    std::span<const StokesPixel> reference) {

    if (image.empty() || image.size() != reference.size()) {
        throw std::invalid_argument(
            "Image error requires non-empty Stokes images with equal sizes.");
    }
    require_finite_stokes(reference, "reference");
    require_finite_stokes(image, "comparison image");
    double denominator = 0.0;
    for (const StokesPixel& pixel : reference) {
        denominator += std::abs(pixel[0]);
    }
    if (!(denominator > 0.0) || !std::isfinite(denominator)) {
        throw std::domain_error(
            "Image error is undefined for zero or non-finite reference Stokes I.");
    }
    std::array<double, 4> result = {};
    for (size_t pixel = 0; pixel < image.size(); pixel++) {
        for (size_t stokes = 0; stokes < result.size(); stokes++) {
            result[stokes] +=
                std::abs(image[pixel][stokes] - reference[pixel][stokes]);
        }
    }
    for (double& value : result) {
        value /= denominator;
        if (!std::isfinite(value)) {
            throw std::domain_error("Image error result is non-finite.");
        }
    }
    return result;
}

} // namespace grrt
